drivers/crypto/crypto_rts5912_sha.c

```diff
@@ -0,0 +1,323 @@
+	uint32_t remain, be_high, be_low, ret_val = 0;
+
+	k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+	remain = rts5912_sha256_ctx->total[0] & (RTS5912_SHA2DMA_BLOCK_SIZE - 1);
+	rts5912_sha256_ctx->buffer[remain++] = 0x80U;
+
+	if (remain <= RTS5912_SHA2_BLOCK_EXTEND_CHECK) {
+		memset((void *)(&(rts5912_sha256_ctx->buffer[remain])), 0x0,
+		       RTS5912_SHA2_BLOCK_EXTEND_CHECK - remain);
+	} else {
+		memset((void *)(&(rts5912_sha256_ctx->buffer[remain])), 0x0,
+		       RTS5912_SHA2DMA_BLOCK_SIZE - remain);
+		k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+		ret_val = rts5912_sha256_process(dev, (uint8_t *)(&(rts5912_sha256_ctx->buffer[0])),
+						 0x1);
+		if (ret_val != 0) {
+			return ret_val;
+		}
+		k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+		memset((void *)(&(rts5912_sha256_ctx->buffer[0])), 0x0,
+		       RTS5912_SHA2_BLOCK_EXTEND_CHECK);
+	}
+
+	be_high = (rts5912_sha256_ctx->total[0] >> (32 - RTS5912_SHA2DMA_8Byte_SHIFT)) |
+		  (rts5912_sha256_ctx->total[1] << RTS5912_SHA2DMA_8Byte_SHIFT);
+	be_low = rts5912_sha256_ctx->total[0] << RTS5912_SHA2DMA_8Byte_SHIFT;
+	*(uint32_t *)(&(rts5912_sha256_ctx->buffer[56])) = sys_cpu_to_be32(be_high);
+	*(uint32_t *)(&(rts5912_sha256_ctx->buffer[60])) = sys_cpu_to_be32(be_low);
+
+	k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+	ret_val = rts5912_sha256_process(dev, (uint8_t *)(&(rts5912_sha256_ctx->buffer[0])), 0x1);
+	if (ret_val != 0) {
+		return ret_val;
+	}
+	k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+
+	volatile uint32_t *output_buf = (volatile uint32_t *)output;
+
+	for (uint32_t i = 0; i < 7; i++) {
+		output_buf[i] = sys_cpu_to_be32(rts5912_sha256_ctx->state[i]);
+	}
+	if (rts5912_sha256_ctx->is224 == 0) {
+		output_buf[7] = sys_cpu_to_be32(rts5912_sha256_ctx->state[7]);
+	}
+	k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+	return ret_val;
+}
+
+static int rts5912_sha256_handler(struct hash_ctx *ctx, struct hash_pkt *pkt, bool finish)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx =
+		(struct rts5912_sha256_context *)(ctx->device)->data;
+	int ret_val = 0;
+
+	k_mutex_lock(&(rts5912_sha256_ctx->crypto_rts5912_in_use), K_FOREVER);
+	memcpy(rts5912_sha256_ctx->sha2_data_in_sram, pkt->in_buf, pkt->in_len);
+	k_mutex_unlock(&(rts5912_sha256_ctx->crypto_rts5912_in_use));
+	ret_val = rts5912_sha256_update(ctx->device, rts5912_sha256_ctx->sha2_data_in_sram,
+					pkt->in_len);
+
+	if (ret_val) {
+		return ret_val;
+	} else if (finish) {
+		ret_val = rts5912_sha256_finish(ctx->device, pkt->out_buf);
+	}
+
+	return ret_val;
+}
+
+static int rts5912_hash_begin_session(const struct device *dev, struct hash_ctx *ctx,
+				      enum hash_algo algo)
+{
+	struct rts5912_sha256_context *rts5912_sha256_ctx = dev->data;
+
+	if (rts5912_sha256_ctx->in_use == TRUE) {
+		printk("Crypto driver is busy!");
+		return -EBUSY;
+	}
+
+	rts5912_sha256_ctx->in_use = TRUE;
```

Comment:
`in_use` is now part of the `struct rts5912_sha256_context` so it should be protected with the mutex as you did in other occurrences of that structure in this file. You should grab the lock before checking `in_use` at line 248 and release it after you setting it at line 253 (of course you should also release it in case of failure in the `if` statement.